src/state.rs

```diff
@@ -94,14 +110,45 @@ impl MintersStore {
 pub static BALANCES: Item<u128> = Item::new(PREFIX_BALANCES);
 pub struct BalancesStore {}
 impl BalancesStore {
+    fn save(store: &mut dyn Storage, account: &Addr, amount: u128) -> StdResult<()> {
+        let balances = BALANCES.add_suffix(account.as_str().as_bytes());
+        balances.save(store, &amount)
+    }
+
     pub fn load(store: &dyn Storage, account: &Addr) -> u128 {
         let balances = BALANCES.add_suffix(account.as_str().as_bytes());
         balances.load(store).unwrap_or_default()
     }
 
-    pub fn save(store: &mut dyn Storage, account: &Addr, amount: u128) -> StdResult<()> {
-        let balances = BALANCES.add_suffix(account.as_str().as_bytes());
-        balances.save(store, &amount)
+    pub fn update_balance(
+        store: &mut dyn Storage,
+        account: &Addr,
+        amount: u128,
+        decoys: Option<Vec<Addr>>,
+        entropy: Option<Binary>,
+    ) -> StdResult<()> {
+        match decoys {
+            None => Self::save(store, account, amount),
+            Some(decoys_vec) => {
+                let mut accounts_to_be_written: Vec<(Addr, u128)> = vec![];
+
+                accounts_to_be_written.push((account.clone(), amount));
+                for decoy in decoys_vec.iter() {
+                    // Please note that decoys are not always present in the DB. In this case it is ok beacuse load will return 0.
+                    accounts_to_be_written.push((decoy.clone(), Self::load(store, decoy)));
```

Comment:
if you load the previous decoy balance in a line of its own, you can save the `clone()`, (probably)